invertedai_cpp/invertedai/logger.cc

```diff
@@ -0,0 +1,131 @@
+        UL time_ms = this->get_current_time_in_milliseconds_();
+
+        if (mode == "location_info"){
+            this->loc_request_times_.push_back(time_ms);
+            this->loc_requests_.push_back(req);
+        }
+        else if (mode == "initialize"){
+            this->init_request_times_.push_back(time_ms);
+            this->init_requests_.push_back(req);
+        }
+        else if (mode == "drive"){
+            this->drive_request_times_.push_back(time_ms);
+            this->drive_requests_.push_back(req);
+        }
+    };
+
+    void LogWriter::append_response(const std::string &res, const std::string &mode){
+        UL time_ms = this->get_current_time_in_milliseconds_();
+
+        if (mode == "location_info"){
+            this->loc_response_times_.push_back(time_ms);
+            this->loc_responses_.push_back(res);
+        }
+        else if (mode == "initialize"){
+            this->init_response_times_.push_back(time_ms);
+            this->init_responses_.push_back(res);
+        }
+        else if (mode == "drive"){
+            this->drive_response_times_.push_back(time_ms);
+            this->drive_responses_.push_back(res);
+        }
+    };
+
+    void LogWriter::write_log_to_file(const std::string &dir_path){
+        json log;
+
+        // log["location_requests"] = this->write_iai_data_vector_to_log_<LocationInfoRequest>(this->loc_requests_);
+        // log["location_responses"] = this->write_iai_data_vector_to_log_<LocationInfoResponse>(this->loc_responses_);
+
+        log["location_requests"] = this->loc_requests_;
+        log["location_responses"] = this->loc_responses_;
+
+        log["location_request_times"] = this->loc_request_times_;
+        log["location_response_times"] = this->loc_response_times_;
+
+        // log["initialize_requests"] = this->write_iai_data_vector_to_log_<InitializeRequest>(this->init_requests_);
+        // log["initialize_responses"] = this->write_iai_data_vector_to_log_<InitializeResponse>(this->init_responses_);
+
+        log["initialize_requests"] = this->init_requests_;
+        log["initialize_responses"] = this->init_responses_;
+
+        log["initialize_request_times"] = this->init_request_times_;
+        log["initialize_response_times"] = this->init_response_times_;
+
+        // log["drive_requests"] = this->write_iai_data_vector_to_log_<DriveRequest>(this->drive_requests_);
+        // log["drive_responses"] = this->write_iai_data_vector_to_log_<DriveResponse>(this->drive_responses_);
+
+        log["drive_requests"] = this->drive_requests_;
+        log["drive_responses"] = this->drive_responses_;
+
+        log["drive_request_times"] = this->drive_request_times_;
+        log["drive_response_times"] = this->drive_response_times_;
+
+        std::string file_path = "iai_log" + std::to_string(this->get_current_time_in_milliseconds_()) + ".json";
+        std::string full_path = dir_path + file_path;
```

Comment:
If `dir_path` does not end with a slash will this work ?

Reply:
Unfortunately it will "work" but everything before the last slash will be added to the name of the file instead of the directory path.

Reply:
This issue I think will be resolved by following Adam's suggestion of trying to make the directory if it doesn't exist.

Reply:
Sure as long as you grabbed the correct name with path when it's created.  Or use something like `boost::filesystem::path` it should handle both cases

Reply:
So the functions in the std::filesystem library didn't pad the path with a "/" as expected so I added a line to check if it ended in a "/" and add one if it is not present.